File: tc-arg/contest-09/k_eugeny_and_play_list.py
# TCArg 2020 - Contest #9
# Eugeny and Play List
# https://codeforces.com/group/j1UosVRZar/contest/288666/problem/K

def find_song(play_list, moment):
	n = len(play_list)
	left = 0
	right = n - 1
	result = -1
	while left <= right:
		mid = (left + right) // 2
		##  print(f"left={left}, mid={mid}, right={right}, moment={moment}")
		if play_list[mid][1] <= moment <= play_list[mid][2]:
			result = play_list[mid][0]
			break
		elif moment < play_list[mid][1]:
			right = mid - 1
		else:
			left = mid + 1
	return result


if __name__ == "__main__":
	n, m = [int(x) for x in input().split()]
	play_list = []
	current_time = 1
	for i in range(n):
		ci, ti = [int(x) for x in input().split()]
		start_time = current_time
		end_time = start_time + ci * ti - 1
		play_list.append((i + 1, start_time, end_time))
		current_time = end_time + 1
	moments = [int(x) for x in input().split()]
	for m in moments:
		print(find_song(play_list, m))
		

File: tc-arg/contest-09/k_eugeny_and_play_list.cpp
// TCArg 2020 - Contest #9
// Eugeny and Play List
// https://codeforces.com/group/j1UosVRZar/contest/288666/problem/K

#include <iostream>
#include <vector>
#include <tuple>

using namespace std;

int find_song(vector<tuple<int, int, int>>& play_list, int moment)
{
	int n = play_list.size();
	int left = 0;
	int right = n - 1;
	int result = -1;

	while (left <= right)	{
		int mid = left + (right - left) / 2;
		if (get<1>(play_list[mid]) <= moment && moment <= get<2>(play_list[mid]))	{
			result = get<0>(play_list[mid]);
			break;
		}
		else if (moment < get<1>(play_list[mid]))	{
			right = mid - 1;
		}
		else	{
			left = mid + 1;
		}
	}
	return result;
}

int main()
{
	int n, m;
	vector<tuple<int, int, int>> play_list;
	int ci, ti, vi, current_time, start_time, end_time;

	ios::sync_with_stdio(0);
	cin.tie(0);

	cin >> n >> m;
	current_time = 1;
	for (int i = 0; i < n; i++)	{
		cin >> ci >> ti;
		start_time = current_time;
		end_time = start_time + ci * ti - 1;
		play_list.push_back(make_tuple(i + 1, start_time, end_time));
		current_time = end_time + 1;
	}
	for (int i = 0; i < m; i++)	{
		cin >> vi;
		cout << find_song(play_list, vi) << "\n";
	}
	return 0;
}


